Add ConsoleExternalLogLink type

jhadvig opened the pull request:
Story: https://jira.coreos.com/browse/CONSOLE-1546

Based on the changes in https://github.com/openshift/console-operator/pull/246/files

Adding comments to open discussion regarding the naming. Will add generated code afterwards. 

/assign @spadgett @benjaminapetersen

@rhamilto  fyi

jhadvig commented on console/v1/types.go:
```diff
@@ -7,3 +7,22 @@ type Link struct {
 }
+
+//  Represents ability to provide custom pod log links.
+type LogLink struct {
```
Wasn't really sure if w want to have it in the `types.go` or directly `types_console_log_links.go`, but since it was similar to the `Link` struct, I've put it here.

jhadvig replied:
Or maybe:
- `LogLinkTemplate`
- `PodLogLink`
- `PodLogLinkTemplate`

Wasn't sure regarding the name here..

benjaminapetersen replied:
I put `Link` here as it is shared among several of the other resources.  I don't think this one will be shared, so it may be best in `types_console_log_links.go`.

The naming is tricky agree. Since this one is embedded, it doesn't matter quite as much, but it would be good if it was clear.

spadgett replied:
This will eventually be generic for other types, so I would avoid using "pod" in the name.

benjaminapetersen commented on console/v1/register.go:
```diff
@@ -32,6 +32,7 @@ func Resource(resource string) schema.GroupResource {
 func addKnownTypes(scheme *runtime.Scheme) error {
 	scheme.AddKnownTypes(GroupVersion,
 		&ConsoleLink{},
+		&ConsoleLogLink{},
```
I mentioned on slack, not sure I'm happy with `ConsoleLink` and `ConsoleLogLink`.  The names are very similar, and the word `Log` doesn't really clarify intent.

The new one is something more like `ConsoleExternalServiceLink`. (`Service` may also be an overloaded term, however).  `ConsoleLink` is really for simple menu links.  The new link is for much more contextual link generation that could be for logs, monitoring, probably other kinds of services.

spadgett replied:
This is so specific to logs, I'm not sure "External Service Link" is good. Do we have any other specific kinds of external services we're planning to add? Without knowing what will come later, it's difficult to say whether the properties here will work for other kinds of services. It's going to be hard to add a different kind of external service later if we need completely different properties.

I agree the names are too close. Maybe `ConsoleExternalLogLink` is better?

benjaminapetersen replied:
From `resourceName`, `containerName`, `resourceNamespace`, `podLabels` it seems like this is pretty generic.  If we had another use case arise, would we expand this further, or would we rather create a third CRD?  I'd lean towards just expanding available parameters here. Otherwise, we would end up with `ConsoleLink`, `ConsoleLogLink`, `ConsoleFooLink`, seems like potential for more confusion.

Put another way, do we want to make it clear that we do not desire to add any other resources, but would rather expand this one (if future additional needs arise)?

spadgett commented:
cc @bparees

spadgett commented on console/v1/types.go:
```diff
@@ -7,3 +7,19 @@ type Link struct {
+	// - podLabels
+	//
+	// e.g., https://example.com/logs?resourceName=\${resourceName}&containerName=\${containerName}&resourceNamespace=\${resourceNamespace}&podLabels=\${podLabels}
+	HrefTemplate string `json:"href"`
```
```suggestion
	HrefTemplate string `json:"hrefTemplate"`
```

spadgett commented on console/v1/types.go:
```diff
@@ -7,3 +7,19 @@ type Link struct {
+	// text is the display text for the link
+	Text string `json:"text"`
+	// Absolute secure URL (must use https) for the log link including variables
+	// to be replaced. Supported variables, which must exactly match, include:
```
We need to be more explicit about the exact format of the variables in the template. The example hints at it, but better to make it very clear.

spadgett commented on console/v1/types.go:
```diff
@@ -7,3 +7,19 @@ type Link struct {
 }
+
+//  Represents ability to provide custom pod log links.
+type LogLink struct {
```
If this isn't being reused, I'd just inline it in `ConsoleLogLinkSpec`

spadgett commented on console/v1/types_console_log_links.go:
```diff
@@ -0,0 +1,23 @@
+type ConsoleLogLinkSpec struct {
+	LogLink `json:",inline"`
+	// An optional, regular expression used to restrict a log link to a matching
+	// namespace (e.g., ^openshift-).
```
You should mention that the links will be displayed for all namespaces if not specified

spadgett replied:
Doc string should start with the JSON field name

spadgett commented on console/v1/types.go:
```diff
@@ -7,3 +7,19 @@ type Link struct {
+type LogLink struct {
+	// text is the display text for the link
+	Text string `json:"text"`
+	// Absolute secure URL (must use https) for the log link including variables
```
Doc string should start with the JSON field name

spadgett commented on console/v1/types.go:
```diff
@@ -7,3 +7,19 @@ type Link struct {
 	Href string `json:"href"`
 }
+
+//  Represents ability to provide custom pod log links.
```
Doc string should start with the struct name

jhadvig commented:
@spadgett @benjaminapetersen I've udpated the PR addressing your comments.
Went with `ConsoleExternalLogLink` as a alternative to `ConsoleLogLink`, but I think now we will need to update both https://github.com/openshift/console/pull/1892/files and https://github.com/openshift/console-operator/pull/246 with that name, right ?

benjaminapetersen commented:
@jhadvig yup, whatever we land on has to match across these PRs.

bparees commented on console/v1/types_console_log_links.go:
```diff
@@ -0,0 +1,38 @@
+	// matching namespace (e.g., ^openshift-). If not specified, links will be
+	// displayed for all the namespaces.
+	// + optional
+	NamespaceFilter string `json:"namespaceFilter,omitempty"`
```
Do i always get the k8s log link, and then if the namespacefilter matches, i get an additional (external) log link?

What happens if i have multiple ConsoleExternalLogLink objects, for an overlapping set of namespacefilters?

jhadvig replied:
> Do i always get the k8s log link, and then if the namespacefilter matches, i get an additional (external) log link?

I believe that you only get the link that  matches the `namespaceFilter`

> What happens if i have multiple ConsoleExternalLogLink objects, for an overlapping set of namespacefilters?

Not really sure with this one, @rhamilto cau you assist on this one ?

rhamilto replied:
> What happens if i have multiple ConsoleExternalLogLink objects, for an overlapping set of namespacefilters?

You get multiple links.  Note the filter is optional.  You'd get as many links as you have instances that aren't filtered in addition to any instances that match the filter.

jhadvig replied:
@bparees ˆˆ

bparees replied:
ok.  i'm a little surprised we remove the k8s log link when a custom log link exists, seems like that could still be useful? 

But i assume you guys have been through this w/ users/customers/use-cases?

spadgett replied:
@bparees This link is on the pod log page. So you're already viewing the pod log, but have a link out to Kibana or another external logging solution. Bascially it's parity with what we have in 3.x:

https://docs.okd.io/3.11/install_config/web_console_customization.html#extension-option-for-external-logging-solutions

@jhadvig I think we should make it more clear where the log link will appear.

bparees replied:
got it.  sgtm.

spadgett commented on console/v1/types_console_external_log_links.go:
```diff
@@ -0,0 +1,40 @@
+	// text is the display text for the link
+	Text string `json:"text"`
+	// hrefTemplate is an absolute secure URL (must use https) for the log link including
+	// variables to be replaced. Supported variables, which must exactly match, include:
```
We should say something like:

> Variables are specified in the URL with the format `${variableName}`, for intsance, `${containerName}`.

I'd probably take out `which must exactly match`. That seems self-evident.

spadgett commented on console/v1/types_console_external_log_links.go:
```diff
@@ -0,0 +1,40 @@
+	// - podLabels - JSON representation of labels matching the pod with the logs
+	//             - e.g. `{"key1":"value1","key2":"value2"}`
+	//
+	// Resource might be a pod, build, buildconfig, deployment or deploymentconfig.
```
For now, resource is only pods.

spadgett commented on console/v1/types_console_external_log_links.go:
```diff
@@ -0,0 +1,40 @@
+	Spec              ConsoleExternalLogLinkSpec `json:"spec"`
+}
+
+// ConsoleExternalLogLinkSpec is the desired console pod log link configuration.
```
I would state the link will appear on the logs tab of the pod details page.

benjaminapetersen replied:
Something like:`ConsoleExternalLogLinkSpec is the desired log link configuration.  The log link will appear on the logs tab of the pod detail page.`

benjaminapetersen commented on console/v1/types_console_external_log_links.go:
```diff
@@ -0,0 +1,40 @@
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ConsoleExternalLogLink is an extension for customizing OpenShift web console pod log links.
```
can prob drop `pod` from `customizing OpenShift web console pod log...`

benjaminapetersen commented on console/v1/types_console_external_log_links.go:
```diff
@@ -0,0 +1,40 @@
+	Spec              ConsoleExternalLogLinkSpec `json:"spec"`
+}
+
+// ConsoleExternalLogLinkSpec is the desired console pod log link configuration.
```
Can prob drop `pod` from `...desired console pod log...` since we renamed.

benjaminapetersen commented on console/v1/types_console_external_log_links.go:
```diff
@@ -0,0 +1,40 @@
+	Text string `json:"text"`
+	// hrefTemplate is an absolute secure URL (must use https) for the log link including
+	// variables to be replaced. Supported variables, which must exactly match, include:
+	// - resourceName - name of the resource which containes the logs
```
We should probably include the `${ }` syntax in each of these:

```
// - ${resourceName} - name of the resource which containes the logs
// ...
```
To make it explicit (just like the example).

benjaminapetersen commented on console/v1/types_console_external_log_links.go:
```diff
@@ -0,0 +1,40 @@
+	// e.g., https://example.com/logs?resourceName=${resourceName}&containerName=${containerName}&resourceNamespace=${resourceNamespace}&podLabels=${podLabels}
+	HrefTemplate string `json:"hrefTemplate"`
+	// namespaceFilter is a regular expression used to restrict a log link to a
+	// matching namespace (e.g., ^openshift-). If not specified, links will be
```
`...to a matching set of namespaces...`, it may be a set of one, or many.

benjaminapetersen replied:
I think the example has to be:
  `/^openshift-/g`
If you just provide `^openshift-`, our code isn't going to treat that like a regex 

https://github.com/openshift/console/pull/1892/files#r304608136

spadgett replied:
@benjaminapetersen @jhadvig I don't think this comment is right... It will be treated a string:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/search

> A regular expression object. If a non-RegExp object obj is passed, it is implicitly converted to a RegExp by using new RegExp(obj).

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp

benjaminapetersen replied:
My mistake! We can revise that.

jhadvig commented:
@spadgett @benjaminapetersen thanks for the review and suggestions ! Much appreciated.
PTAL

bparees commented:
/approve

spadgett commented:
/hold

spadgett commented:
We should open a follow on PR to address https://github.com/openshift/api/pull/380#discussion_r304619157 Would be good to test with @rhamilto's changes to verify

spadgett commented:
Yeah, confirmed the original example works

```
"openshift-console".search("^openshift-")
0
```

If we wanted to be really clear, we could say it's a string converted into a regular expression using the JavaScript `RegExp` constructor.

benjaminapetersen commented:
Tested a handful: 

```javascript
var str = "openshift-is-cool-openshift-is-fun"

var regex1 = "openshift-"
var regex2 = "^openshift-"
var regex3 = "/^openshift-/g"
var regex4 = /^openshift-/g

console.log("regex1", str.search(regex1)); // returns index 0
console.log("regex2", str.search(regex2)); // returns index 0
console.log("regex3", str.search(regex3)); // returns index -1
console.log("regex4", str.search(regex4)); // returns index 0
```
So since this is coming from `yaml`, as a `string`, these will be quote wrapped when they land in JS.  Thus `regex3` ` "/^openshift-/g"` doesn't work.
